[1주차] javaScript 숫자 야구게임 미션 제출합니다.

dkr-sjr opened the pull request:
안녕하세요, 김범수 리뷰어님(@Indigochi1d ).
저는 이번에 그리디 3기로 활동하게 된 강건 이라고 합니다. 앞으로 한 한기 동안 잘 부탁 드리겠습니다!
이번 미션으로 제가 처음으로 html과 javaScript를 다뤄보게 되었습니다. 이전에는 학교에서 배운 c 언어와 독학으로 c++를 조금 공부해서 사용해 왔었습니다. javaScript를 다룬것이 처음이라 부족한 점이 아주 많을 것 같습니다. 그래도 열정을 가지고 배워보겠습니다!
바쁘신 와중에도 시간을 내어 리뷰를 해주셔서 감사합니다!

---
### 필요 기능
1. 페이지 초기화
2. 게임 실행
3. 결과 출력
4. 재시작

위 기능이 제가 생각한 이번 미션의 핵심 기능입니다.
가장 먼저 위 기능들을 구현하기 위해  어떤 클래스가 필요할지 생각해 보았습니다.
처음에는 게임 로직을 담당 할 BaseballGame 클래스와 웹과 BaseballGame 클래스 사이에서 중간 매개를 해줄 클래스, 이렇게 2개의 클래스를 만들어서 사용하려고 하였으나, dom의 기능을 사용해서 특정 id를 가진 태그에 이벤트를 활당하면 중간매개 클래스가 불필요 하다고 생각이 들어 최종적으론 하나의 클래스를 사용하였습니다. 
그 외에 요구 사항을 지키기 위해 BaseBallGame 클래스 외에 checkInput, printResult, GetRandomNumbers 이렇게 3개의 함수를 추가로 작성하였습니다. 해당 함수들을 BaseballGame 클래스의 메서드로 선언 할지 고민해 봤지만, 해당 함수가 하는 일이 BaseballGame 과는 어울리지 않다고 생각이 들어 따로 함수로 만들어 사용하였습니다.

<img width="1934" height="1558" alt="image" src="https://github.com/user-attachments/assets/f82d4aea-336c-4e70-abbd-25f9dccf3326" />
테스트 결과입니다!

---
### 질문사항
저는 코드를  Baseballgame 클래스를 담은 파일, 추가로 필요한 함수들을 담은 파일, html에서 불러올 파일 이렇게 나누어 작성하였습니다. 그중에서 index.js에서 다른 언어들과 다르게 main과 같은 함수가 아닌 그냥 파일 자체에 바로 코드를 작성 하였습니다.
이 방식이 옳은 방식인지 궁금합니다. 제가 찾아본 바로는 이렇게 작성해도 된다고 하는데, 다른 맞는 방법이 있는지 알고 싶습니다.

저는 BaseballGame에 포함되지 않은 함수들을 따로 모아서 선언하는 방식을 사용하였는데, 이렇게 모듈화를 해도 될지는 의문이 듭니다. 이렇게 모듈화를 할 예정인 함수들을 따로 모아서 클래스로 만들어야 할까요? c++에선 함수를 따로 만들지 않고 클래스 메서드만 사용하다 보니 헷갈립니다. 

코드를 작성하기 전에 기능 목록을 작성한 후에 그 기능들을 순서대로 작성하며 커밋하라는 요구사항이 있었습니다.
처음 제 나름대로 기능 목록을 작성한 후에 진행하였지만, 진행 도중 계획과는 맞지 않거나 변경해야할 사항들이 생겨 중간에 기능 목록을 몇 번 수정하곤 했습니다. 제가 궁금한 점은 처음 코드를 작성하기 전에 기능 목록을 어느 정도로 작성해 놓아야 할지 궁금합니다. 기능 목록을 작성하는데 좀 더 많은 시간을 투자하는게 좋을까요? 기능 목록 작성 요령이 궁금합니다.

마지막으로 리뷰어님은 필요한 기능이나 함수가 있을 때 어떠한 방법으로 찾아서 사용하시는지도 궁금합니다!

제가 처음으로 pull request를 작성하다 보니 두서 없이 작성하게 된 것 같습니다. 제가 잘못한 부분이나 고쳐야 하는 부분이 있다면 가감 없이 말해주시면 감사하겠습니다!

dkr-sjr committed:
docs(README) : 기능 목록 작성

요구 사항에 따라 기능 목록을 작성함.
나중에 변경될 수 있음.

dkr-sjr committed:
feat(index.js) : index.js 파일 생성

src 파일에 index.js 파일 생성함.
내부에 BaseBallGame 클래스 작성함.
생성자와 초기화 함수 작성함.

docs(readme) : 생성자가 필요할것 같아 기능 목록에 추가함.

dkr-sjr committed:
docs(readme) : 기능 목록 수정

dom과 event에 대해 알아보니
중간 매개 클래스가 필요하지 않을것 같아 삭제.
그에 맞게 BaseBallGame 클래스 재작성

dkr-sjr committed:
docs(readme) : 기능 목록 수정

프로그래밍 요구사항에 맞게 baseballgame의 역할을 축소.
기능 목록을 좀 더 단계적으로 나눔.

dkr-sjr committed:
feat(index.js) : compare 함수 구현

play 함수를 구현하기 위해 그 안에서 사용할 compare 함수를 구현함.

dkr-sjr committed:
feat(index.js) : play 함수 구현

이전에 작성한 compare 함수를 이용하여 play 함수 구현 완료.
indent depth가 2를 넘기지 않기 위해 함수를 분리해서 작성함.

dkr-sjr committed:
feat(index.html) : 첫 초기화 구현

페이지에 방문 했을 때 정답 값을 랜덤 값으로 설정하는 로직 작성.

dkr-sjr committed:
feat(idex.js) : 확인 버튼 이벤트 구현

html의 확인 버튼에 이벤트를 활당함.
버튼 클릭시 텍스트에서 값을 읽어와 저장된 랜덤 값과 비교후 해당 결과를 저장.

fix(index.html) : js 코드를 html에서 script 태그 안에서 실행하던것을 index.js 파일 안으로 이동함.

dkr-sjr committed:
fix(index.js) : querySelector 사용

기존에 getElementById를 querySelector 로 변경함.

dkr-sjr committed:
feat(index.js) : checkInput 함수 추가

공통 요구 사항을 지키기 위하여 입력 값 검사 기능을 함수로 분리함.

docs(readme) : 필요 메서드 추가

dkr-sjr committed:
fix(index.js) : checkInput 함수가 수정

checkInput 함수가 제대로 작동하지 않던 것을 수정함.

dkr-sjr committed:
feat(index.js) : printResult 함수 추가

비교 결과로 얻은 결과를 출력하는 함수를 추가함.
결과에 따라 재시작 버튼이 보이고 안 보이게 변경함

dkr-sjr committed:
feat(index.js) : getRandomNumbers 함수 추가

중복 코드 사용을 막기 위해 기존의 초기화 로직을 함수화 하여 재사용함.
이를 재시작 버튼에 이벤트 함수에서 호출함.

dkr-sjr committed:
feat(index.js) : makeResult 함수 추가

기존에 play에 있던 로직을 makeResult로 분리함.

fix(index.js) : 전체적으로 요구사항에 맞게 수정함.

dkr-sjr committed:
fix(src) : 코드 모듈화

모듈화 요구사항을 지키기 위하여
index.js에 있던 코드를 분리함.

Indigochi1d commented on README.md:
```diff
@@ -8,7 +8,26 @@
+2. 확인 버튼 클릭 이벤트 - 버튼 클릭시 userinput에서 값을 가져와 BaseBallGame의 play 함수를 이용하여 정답과 비교
+3. play 함수의 반환 값에 따른 출력 - 반환 값에 따라 어떤것을 출력 할지 결정
+4. 게임 재시작 기능 - 정답을 맞췄다면 정답 값을 다시 랜덤 설정 
+
```
👍🏻

Indigochi1d commented on cypress/fixtures/example.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Using fixtures to represent data",
+  "email": "hello@cypress.io",
+  "body": "Fixtures are a great way to mock data for responses to routes"
+}
```
이 파일은 무엇인가요?

dkr-sjr replied:
해당 파일은 npm install 과정에서 생긴 파일 같습니다. 제가 작성한 파일은 아니고 테스트를 하는 과정에서 추가된 파일입니다!

Indigochi1d commented on index.html:
```diff
@@ -17,14 +17,14 @@ <h1>⚾ 숫자 야구 게임</h1>
   </div>
   <script src="https://cdn.jsdelivr.net/npm/@woowacourse/mission-utils@1.0.1/dist/mission-utils.min.js"></script>
-  <script type="module" src="src/index.js"></script>
+  <script type="module" src="src/index.js" defer></script>
```
`defer` 을 넣으실 때 어떤 역할을 기대하고 넣으신건가요?

dkr-sjr replied:
html 을 읽어 내리는 도중에 <script> 태그를 만나면 html 파일을 읽던 것을 중단하고 해당 스크립트를 실행하고 다시 html를 읽어 내려갑니다.  이때 defer를 사용해 준다면 html를 읽던 도중에 멈추지 않고 다 읽은 후에 <script>를 실행해 준다고 합니다. 이를 통해 웹 로딩을 빠르게 하고 혹여나 발생할 에러를 방지해 준다고 배웠습니다. 

근데 알아보니 type = "module" 가 defer를 사용한 것처럼 작동하기에 defer 은 붙이지 않았어도 상관 없는 것 같습니다!
또한 script 태그가 에초에 html의 마지막에 붙어 있기에 에러가 날 일도 없었을 것 같습니다.

Indigochi1d commented on src/BaseBallGame.js:
```diff
@@ -0,0 +1,44 @@
+    let ballCount = 0;
+
+    for(let i = 0 ; i < 3; i+=1){
+      let result = this.compare(computerInputNumbers,userInputNumbers[i], i);
```
`result`를 `let`으로 선언한 이유가 있으신가요? `let`과 `const`의 차이에 대해서 건님이 학습하신 것을 공유해주실 수 있나요?

dkr-sjr replied:
const 의 역할은 const로 선언한 변수에 값을 재할당을 금지해주는 역할을 합니다. 따라서 나중에 다시 변경할 일이 없는 변수에 const를 붙여줌으로서 혹여나 발생할 실수를 방지해 줍니다.
let은 const 와 반대로 재할당이 필요한 변수를 선언할 때 사용해 줘야 합니다.
var 대신 let과 const를 사용하는 이유는 var는 함수 스코프이며 const와 let은 블럭 단위 스코프 입니다. 따라서 {} 안에서 선언한 변수는 {} 밖에선 접근 불가능 합니다. 지역 변수와 같은 역할을 하게 됩니다.

제가 작성한 코드를 다시 보니 result를 let으로 선언할 이유가 없는 것 같습니다. const가 더 적합합니다!

Indigochi1d commented on src/BaseBallGame.js:
```diff
@@ -0,0 +1,44 @@
+    if(strikeCount !==0){
+      output += (strikeCount.toString() +'스트라이크')
+    }
+    if(output ==='')
+    {
+      output = '낫싱'
+    }
```
사소하지만 코드 컨벤션을 지키려는 노력은 굉장히 중요해요. 
- 세미콜론이 없어요.
- brace의 위치가 일정하지 않아요

> 공통 요구사항
> [Airbnb 자바스크립트 코드 컨벤션](https://github.com/airbnb/javascript)을 지키면서 프로그래밍 한다.
```suggestion
    if(output ===''){
      output = '낫싱';
    }
```

나중에 어떠한 개발직군이든 집단내의 코드 품질을 보장하며 코드를 작성하는 것은 매우 중요한 일이에요! 그렇기 위해서 우리는 코드를 쓰는 규칙과 약속을 정하는데 그것을 우리는 **코드 컨벤션**이라고 부릅니다. 문서에 기재된 에어비앤비 코드 컨벤션도 그러한 것들 중 하나에요. 
하지만 실제로 에어비앤비 코드 컨벤션에 적힌 모든 것을 전부 읽고 외운 후에 코드를 작성하는 것은 우리가 컴퓨터가 아니라면 불가능합니다. 그러면 컴퓨터한테 이 코드가 약속을 지킨 코드인지 검사하게 할 수 있는 방법이 있지 않을까요? 한번 찾아보시고 적용해보시면 좋을 것 같습니다!

dkr-sjr replied:
제 나름 코드 컨벤션을 지키려고 해봤는데 실수가 있었던 것 같습니다. 다음에는 좀 더 노력해 보겠습니다!

찾아보니 검사를 할때는 ESlint를 많이 사용하고 추가로 prettier를 통해서 자동으로 코드를 수정한다고 합니다.
저는 아직 배우는 과정이니 검사를 위해 ESlint만 사용하는 것이 좋을까요? 

ESlint를 사용하니 코드 전체가 빨간 줄이 되었습니다..!
대부분 잘 지켰다고 생각했는데, 엄청나게 많은 문제점이 있었네요.

Indigochi1d commented on src/additionalFuctions.js:
```diff
@@ -0,0 +1,37 @@
+  if(!Number.isInteger(digitNumber) || 111 > digitNumber || digitNumber > 999){
+    return false;
+  }
+  let set = [];
```
코드를 읽었을 때 `set`이 무엇임을 알 수 있지만 그저 이 변수명만 보았을 때 무엇을 위한 set인지 파악할 수 있다면 더 좋을 것 같아요! 만약 저라면 `set`보다는 `usedDigits` 라던지 `usedDigitsSet`이라던지로 할 것 같아요. 

그저 set으로 쓰는 것은 자칫 이 변수가[ Set 자료형](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Set)인가? 라고 오해할 수 있기 때문에 변수명을 잘 적는 것은 매우 중요해요.

dkr-sjr replied:
옳은 말씀이십니다. 다음부턴 변수 이름을 지을 때 파악하기 쉬운 이름을 생각해 보겠습니다!
그리고 변수가 배열이라서 let 보단 const가 적절하군요!

Indigochi1d commented on src/additionalFuctions.js:
```diff
@@ -0,0 +1,37 @@
+    document.querySelector('#game-restart-button').style.display = 'block';
+  }
+  document.querySelector('#result').textContent = stringToPrint;
+};
```
파라미터를 재할당하고 있네요! 제 생각에는 함수 파라미터를 직접 수정하는 것은 좋지 않은 패턴인 것 같은데 혹시 건님께서 의도하신 부분이 있을까요?

dkr-sjr replied:
stringToPrint 라는 변수가 화면에 출력하는 스트링을 담기 위한 변수이며 
if(stringToPrint === '3스트라이크') 해당 if문이 발동되지 않는다면 그대로 stringToPrint가 담고 있는 문자열을 출력합니다. 
이러한 이유로 stringToPrint 변수를 재 활당하여 사용하였는데, 확실히 파라미터를 재할당하는 것은 문제가 있을 수도 있을것 같습니다. 
중간 매개 변수를 추가하여 수정해 보겠습니다!

Indigochi1d commented on src/index.js:
```diff
@@ -0,0 +1,31 @@
+let answerNumbers = getRandomNumbers();
+document.querySelector('#game-restart-button').style.display = 'none';
+
+console.log(answerNumbers);
```
이후의 다른 미션에서는 `console.log`를 전부 제거하신 후에 PR을 올려주시는게 좋을 것 같아요!

dkr-sjr replied:
넵, 알겠습니다!

Indigochi1d commented on src/BaseBallGame.js:
```diff
@@ -0,0 +1,44 @@
+    return this.makeResult(strikeCount,ballCount);
+  }
+
+  compare(computerNumbers, userNumber, idx){
+    for(let i =0; i < computerNumbers.length ; i+=1){
+      if(computerNumbers[i] === userNumber && i === idx){
+        return 'strike';
+      }
+      if(computerNumbers[i] === userNumber && i !== idx){
+        return 'ball';
+      }
+    }
+    return 'Nothing';
+  }
```
`compare`이라는 함수명 보다 더 구체적인 함수명이 있을 것 같아요. compare은 동사인데 비교하기 위해서 "무엇을" 비교하는지가 들어가면 좋을 것 같습니다! 그러면 코드를 읽는 타인의 입장에서 이 함수명만 보고도 "아! 이건 무엇을 하겠구나" 하고 짐작하고 로직을 유추할 수 있어요.

또 이건 제 성향이긴하지만 `idx`보다는 `position`이 가독성에 도움되는 파라미터명인 것 같습니다  😄 

또한 스트라이크/볼이 없으면 `Nothing`이 반환되는데 이것이 `play` 함수안에서 `result === Nothing` 일 경우 그냥 버리는 로직이 되어버린 것 같아요. 어떤 의도를 가지고 적어주셨는지 공유해주세요!

dkr-sjr replied:
compare 함수가 baseBallGame의 함수이기에 이름을 좀 간단하게 지었는데, 확실히 단순히 compare이란 이름으론 그 역할을 재대로 알려주지 못하는 것 같습니다. compareNumberWithAnswers 라는 이름으로 변경 하였습니다!
변수나 함수의 적절한 이름을 생각하는 것은 쉬운 일은 아닌것 같습니다. 이름을 통해 역할을 알리기 위해서 구체적으로 이름을 짓다 보면 이름이 너무 길어지거나, 조금 짧게 짓게 된다면 그 의미를 전달하기엔 부족해 보입니다. 
이름이 조금 길더라도 확실한 이름을 짓는게 좋을까요?

학교에서 C를 가장 먼저 배운 영향인지 주로 index 줄여서 idx 라는 변수명을 지금까지 사용해 왔던 것 같습니다. 이런 상황에서는 index보다 position이 더 가독성이 도움이 될까요?  리뷰어 님께서는 index라는 표현을 아예 사용하지 않으시는 건지 아니면 특정 기준이 있는 건지 궁금합니다!


compare 함수의 목적은 정답과 입력 값을 비교하여 결과를 출력하는 것입니다. 따라서 함수의 목적에 따르면 어떤 결과이든지 반환 값이 존재해야 된다고 생각하여 nothing를 반환하였습니다. 이후에 play에서는 사실상 ball과 strike의 개수만 필요하기에 반환 값이 nothing인 경우는 체크를 하지 않았습니다. 프로그램 자체에는 필요 없는 반환 값이지만 함수 자체에는 필요하다고 생각하여 남겨두었습니다!

Indigochi1d commented on src/BaseBallGame.js:
```diff
@@ -0,0 +1,44 @@
+      }
+      else if(result === 'ball'){
+        ballCount+=1;
+      }
```
```suggestion
      if(result === 'strike'){
        strikeCount+=1;
      }
      if(result === 'ball'){
        ballCount+=1;
      }
```
이렇게 사용하는 것이 더 좋아보이는데 건님은 어떻게 생각하시나요? 왜 if-else if를 사용하였는지 공유해주시면 같이 이야기 나누어보아도 좋을 것 같아요

dkr-sjr replied:
위 if 문은 결과가 strike 인지 ball인지 아니면 그 외 값 인지에 따라 서로 다른 코드를 실행시켜야 하는 구문입니다. 
else if 대신 if를 사용하여도 코드 작동에는 문제가 없지만, if else를 사용하는 것이 좀 더 의미에 적합하다고 생각하였습니다!

Indigochi1d commented on src/BaseBallGame.js:
```diff
@@ -0,0 +1,44 @@
+
+  makeResult(strikeCount, ballCount){
+    let output = '';
+    if(ballCount !== 0){
+      output += (ballCount.toString() + '볼 '); 
+    }
+    if(strikeCount !==0){
+      output += (strikeCount.toString() +'스트라이크')
+    }
```
템플릿 리터럴을 사용하는 것은 어떨까요?
```suggestion
    if(ballCount !== 0){
      output += `${ballCount}볼 `;
    }
    if(strikeCount !==0){
      output += `${ballCount}스트라이크`;
    }
```

dkr-sjr replied:
이런 기능이 있는지 몰랐습니다! 다음부터 문자열을 생성할 때는 템플릿 리터럴를 사용해서 작성해 보겠습니다! 감사합니다.

Indigochi1d commented on src/additionalFuctions.js:
파일명 오타 수정해주세요!

dkr-sjr replied:
수정했습니다! 알려주셔서 감사합니다.

Indigochi1d commented on src/index.js:
```diff
@@ -0,0 +1,31 @@
+  answerNumbers = getRandomNumbers();
+  document.querySelector('#game-restart-button').style.display = 'none';
+  document.querySelector('#user-input').value = '';
+  console.log(answerNumbers);
```
마찬가지입니다!

dkr-sjr replied:
console.log 말씀 하시는 거라면 제거했습니다!

Indigochi1d commented on src/additionalFuctions.js:
```diff
@@ -0,0 +1,37 @@
+};
+
+export const getRandomNumbers = ()=>{
+  let output = [];
```
`output`을 `let`으로 선언하셨는데 왜 `let`인지 공유해주세요!

dkr-sjr replied:
output 내부에 값을 추가해야 해서 해당 변수를 변경해야 하는 변수라고 생각했었습니다. 
배열에 값을 추가 하는 건 재 할당이 아니란 걸 간과했습니다! 
let 대신 const가 적절해 보입니다.

Indigochi1d commented on src/index.js:
```diff
@@ -0,0 +1,31 @@
+document.querySelector('#game-restart-button').style.display = 'none';
+
+console.log(answerNumbers);
+const GameHandle = new BaseballGame();
```
보통 변수명을 적을 때는 카멜케이스와 스네이크 케이스가 있어요. 그래서 `gameHandle`이라고 하는 것이 더 적절해보여요.

또한 `GameHandle`이라는 변수명 자체가 조금 어색한 것 같습니다! 건님은 어떻게 생각하시는지 생각을 알려주세요!

dkr-sjr replied:
해당 변수가 class 객체를 만드는 것이라, 제가 class의 경우와 헷갈렸습니다! gameHandle이 적절해 보입니다.

최근 언리얼을 공부 중인데 언리얼에 handle이라는 단어가 자주 사용되다보니, handle이라는 단어가 익숙했나 봅니다.
이 상황은 딱히 생성되어 있는 다른 객체를 다루거나 조작하는 변수를 선언하는 것이 아니다 보니 handle이라는 단어는 어색한 것이 맞는 것 같습니다.
해당 변수는 이 프로그램에서 하나만 선언될 변수이니 baseBallGame 이라고 지어도 괜찮을것 같은데 어떻게 생각하시나요?
클래스와 같은 이름은 적절치 않을까요?

Indigochi1d reviewed (changes requested):
안녕하세요 건님! 저는 그리디 프론트엔드 자료를 만들고 1,2기를 리드했었던 김범수라고 합니다! 반가워요 ㅎㅎ 저는 지금 스타트업에서 인턴을 하고 있어요😄 
첫 미션을 수행해보신 소감은 어떠셨나요? 미션이 주어지고 미션을 수행하기 위해 학습하고 공부하셨을 텐데 그 과정 자체에 박수를 보내드리고 싶습니다 👏🏻 

건님께서는 c와 c++을 학습해보셨다고 했는데 Javascript는 어떠셨나요? Javascript는 앞의 두 언어에 비해 많이 자유로운 언어에요. "이게 되는게 맞다고?" 싶은 것들이 되기 때문에 자유롭지만 그만큼 위험할 수도 있는 언어입니다. 또, javascript를 사용한 개발협업에서는 이 함수의 파라미터와 return 값의 타입을 한눈에 보기가 힘들어요. 함수가 길어지면 길어질 수록 더욱 그렇습니다. 이럴 때 유용한 것이 [JsDocs](https://poiemaweb.com/jsdoc-type-hint)라는 것이에요. 적용법은 그렇게 어렵지 않으니 리팩토링 하실 때 이번 미션에 적용해보시면 많은 도움이 될 수 있을 것 같아요.

구현할 기능 목록을 세분화해서 나눠주신 것은 너무 좋았어요. 앞으로 개발하실 때 이렇게 하시면 좋습니다. 다만 한가지 더 중요한 것을 말씀드리면 "추상화"라는 개념을 개발하실 때 코드에 녹여보는 연습을 해보시면 좋아요. 제가 이해하고 있는 추상화란 **복잡한 세부 구현을 감추고 본질적인 의도만 드러내는 과정**을 의미합니다. 
개발을 하다 보면 구체적인 코드나 절차에 파묻히기 쉬운데, 추상화를 통해 우리는 “어떻게”가 아니라 “무엇을” 하는지에 집중할 수 있습니다. 예를 들어 원의 넓이를 계산하는 함수를 작성할 때 단순히 `3.141592`라는 숫자를 사용하는 대신 `Math.PI`를 쓰는 것이 더 추상화된 표현입니다. **구체적인 상수를 숨기고 `원주율`이라는 의미만 드러내기 때문에 코드의 의도가 한눈에 보입니다.** 
프론트엔드에서는 버튼이나 입력창 같은 UI 요소를 추상화하는 경우가 많아요. 스타일을 직접 작성하는 대신 `Button` 컴포넌트에 `variant="primary"` 같은 속성을 부여하면, 내부 구현을 몰라도 버튼이 어떤 성격을 띠는지 바로 이해할 수 있습니다.

### 질문사항에 대한 저의 생각들
>저는 코드를 Baseballgame 클래스를 담은 파일, 추가로 필요한 함수들을 담은 파일, html에서 불러올 파일 이렇게 나누어 작성하였습니다. 그중에서 index.js에서 다른 언어들과 다르게 main과 같은 함수가 아닌 그냥 파일 자체에 바로 코드를 작성 하였습니다.
이 방식이 옳은 방식인지 궁금합니다. 제가 찾아본 바로는 이렇게 작성해도 된다고 하는데, 다른 맞는 방법이 있는지 알고 싶습니다.

**A:** 충분히 괜찮습니다! 이 프로젝트에서는 따로 그것에 관한 요구사항이 없었기에 충분해요. 다만 프로젝트가 커지고 팀 내의 규칙이 생긴다면 지양해야될 패턴일 수도 있습니다. 코드는 수단이지 목적이 되어서는 안되기 때문에 너무 어떠한 패턴에 얽매이지 않으셔도 좋아요. 하지만 사용자들이 많이 사용하는 아키텍처/패턴을 공부할 필요는 있습니다👍🏻

>저는 BaseballGame에 포함되지 않은 함수들을 따로 모아서 선언하는 방식을 사용하였는데, 이렇게 모듈화를 해도 될지는 의문이 듭니다. 이렇게 모듈화를 할 예정인 함수들을 따로 모아서 클래스로 만들어야 할까요? c++에선 함수를 따로 만들지 않고 클래스 메서드만 사용하다 보니 헷갈립니다

**A:** Javascript의 모듈에 대해서 공부해보시면 좋을 것 같아요. 제가 답을 드리는 것보다 javascript에서의 모듈은 무엇인지 학습해보시면 답을 스스로 찾으실 수 있을 것 같습니다. 
[MDN](https://developer.mozilla.org/ko/docs/Web/JavaScript/Guide/Modules)/[NodeJS](https://nodejs.org/api/modules.html)/[Typescript handbook ](https://www.typescriptlang.org/docs/handbook/2/modules.html)

> 코드를 작성하기 전에 기능 목록을 작성한 후에 그 기능들을 순서대로 작성하며 커밋하라는 요구사항이 있었습니다.
처음 제 나름대로 기능 목록을 작성한 후에 진행하였지만, 진행 도중 계획과는 맞지 않거나 변경해야할 사항들이 생겨 중간에 기능 목록을 몇 번 수정하곤 했습니다. 제가 궁금한 점은 처음 코드를 작성하기 전에 기능 목록을 어느 정도로 작성해 놓아야 할지 궁금합니다. 기능 목록을 작성하는데 좀 더 많은 시간을 투자하는게 좋을까요? 기능 목록 작성 요령이 궁금합니다.

**A:** 가장 어려운 질문 중 하나에요 🤔 위의 문단에서 말씀드렸다 싶이 1️⃣ **기능을 추상화**하고 2️⃣ **그 추상화된것을 코드로 녹인 후** 3️⃣ **코드의 추상화의 패턴**을 가져가 보심이 어떨까요?

> 마지막으로 리뷰어님은 필요한 기능이나 함수가 있을 때 어떠한 방법으로 찾아서 사용하시는지도 궁금합니다!
**A:** 필요한 메서드가 있을 때 어떻게 찾아보는지에 대한 질문으로 이해해도 될까요? 😄  만약 맞다면 저는 어떤 자료형에 따른 메서드들을 전부 문서를 통해 읽어보는 편이에요. 
빠르게 구현을 해야하고 익숙하지 않은 언어를 써야할 때는 그냥 ai에게 물어보거나 필요한 것만 골라서 찾아볼 수 있겠지만 학습 중이라면 진득하게 시간을 가지고 문서를 읽는 편입니다. 

그리디 프론트엔드의 첫 미션을 수행하시느라 대단히 수고 많으셨습니다. 리팩토링 후에 디스코드에서 다시 멘션 주세요. 수고하셨습니다!

dkr-sjr committed:
fix : 리뷰 결과에 따른 수정

전체적으로 commit 리뷰에 따라 문제점들을 수정함.
JSDoc 을 추가하여 타입에 강제성을 부여함.
ESlint extension을 사용하여 Airbnb 코드 컨벤션을 강제함.

dkr-sjr commented:
주말에 시간 내어 리뷰해 주셔서 정말 감사합니다! 

---

Javascript는 다른 언어에 비해 확실히 자유로운 것 같습니다. 이 자유로움이 편하기도 하지만 저에겐 조금 불편한 느낌도 있었습니다. 변수를 선언할때 타입을 정하지 않고 사용해도 문제가 없어 작성할 때는 편리하기도 했지만, 반대로 타입이 눈에 바로 보이지 않아 생기는 불편함도 있었던 것 같습니다. 

리뷰어님께서 소개해주신 JsDocs를 통해 타입을 강제할 수 있다는 것을 알게 되었고, 이를 제 코드에 적용해 보았습니다. 확실히 JsDocs를 사용하면 오류가 발생할 확률을 많이 줄이고 편리함을 제공할 것 같습니다. 반면 이럴 거면 처음부터 변수에 타입을 지정하는 편이 좋지 않나 하는 생각도 들었습니다. 자유로움과 동시에 그로 인해 불편함이 생긴다는 것이 아이러니한 것 같습니다.

저는 클래스를 작성할 때 함수의 위치를 많이 신경 쓰는 편입니다. 이 함수가 이 클래스에서 작성되어야 할 함수인지, 다른 클래스에서 처리할 함수가 아닌지 많이 고려하려고 노력합니다. 리뷰어 님이 말씀해주신 추상화 라는 개념에 대해 좀 더 생각해 보아야겠습니다. 

모듈이라는 개념이 익숙하면서도 어색한 것 같습니다. 다른 언어의 라이브러리와 클래스가 조금 섞여 있는 느낌입니다. 제가 클래스에 포함되지 않은 함수들을 모아서 모듈화 한 건 괜찮은 것 같지만 이름은 맞지 않는 것 같습니다. addtionalFunctions 라는 이름은 어떤 기능들을 담고 있는지 파악할 수 없어 적절치 않아 보입니다. 다른 어떤 이름이 적합한지는 좀 더 생각해 보아야겠습니다. BaseballGame 클래스의 개념을 좀 더 넓게 본다면 클래스 메서드로 넣어도 될 것 같기도 합니다.  

---

코멘트에 답변하기 위해 제가 짠 코드도 다시 읽어보고 생각하는 시간을 가져보니 복습도 되고 새로운 것도 알게 되어 정말 유익한 시간이었습니다. 많은 도움을 주셔서 감사합니다!

Indigochi1d reviewed (approved):
고생하셨습니다! 수정하신 코드도 잘 읽어보았어요. 다만 jsdocs를 쓸 때 모든 변수에다가 지정하는 것은 오히려 가독성을 해칠 수 있습니다. 따라서 함수의 파라미터와 리턴값 정도에만 적용해두는 것이 더 좋아보이고 그것이 일반적인 방식이기도 해요.
또, `"이것은 이 타입이에요"` 라고 개발자가 지정할 뿐이지 실제로는 그 타입인지 아닌지 자바스크립트는 알지 못해요. 그래서 현업에서 타입스크립트를 사용하는 이유이기도 합니다. 

저번에 건님께서 `const`와 `let`의 재할당에 대해서 말씀해주셨어요. Javascript에서의 재할당이란 무엇일까요? Javascript의 메모리 구조는 어떻고 자료형들을 어떤 방식으로 저장하고 있을까요? 이것을 알아보시면 JS를 이해하는데 좋을 것 같아요.

이번 미션 수고하셨고 남은 미션에서도 화이팅입니다!
